book/src/rfcs/RFC0006-Dynamic-Databases.md

```diff
@@ -0,0 +1,635 @@
+    * `DatabaseKeyIndex` values will leak, as we propose no means to reclaim
+      them. However, the same is true of `Slot` values today.
+    * Storing values for the tables directly in the hashtable makes it less obvious how we would return references to them in a safe fashion (before, I had planned to have a separate module that held onto the Arc for the slot, so we were sure the value would not be deallocated; one can still imagine supporting this feature, but it would require some fancier unsafe code reasoning, although it would be more efficient.)
+
+## Motivation
+
+Under the current salsa setup, all of the "glue code" that manages cache
+invalidation and other logic is ultimately parameterized by a type `DB` that
+refers to the full database. The problem is that, if you consider a typical
+salsa crate graph, the actual value for that type is not available until the
+final database crate is compiled:
+
+```mermaid
+graph TD;
+  Database["Crate that defines the database"];
+  QueryGroupA["Crate with query group A"];
+  QueryGroupB["Crate with query group B"];
+  SalsaCrate["the `salsa` crate"];
+  Database -- depends on --> QueryGroupA;
+  Database -- depends on --> QueryGroupB;
+  QueryGroupA -- depends on --> SalsaCrate;
+  QueryGroupB -- depends on --> SalsaCrate;
+```
+
+The result is that we do not actually compile a good part of the code from
+`QueryGroupA` or `QueryGroupB` until we build the final database crate.
+
+### What you can do today: dyn traits
+
+What you can do today is to use define a "dyn-compatible" query group
+trait and then write your derived functions using a `dyn` type as the
+argument:
+
+```rust,ignore
+#[salsa::query_group(QueryGroupAStorage)]
+trait QueryGroupA {
+    fn derived(&self, key: usize) -> usize;
+}
+
+fn derived(db: &dyn QueryGroupA, key: usize) -> usize {
+    key * 2
+}
+```
+
+This has the benefit that the `derived` function is not generic. However, it's
+still true that the glue code salsa makes will be generic over a `DB` type --
+this includes the impl of `QueryGroupA` but also the `Slot` and other machinery.
+This means that even if the only change is to query group B, in a different
+crate, the glue code for query group A ultimately has to be recompiled whenever
+the `Database` crate is rebuilt (though incremental compilation may help here).
+Moreover, as reported in [salsa-rs/salsa#220], measurements of rust-analyzer
+suggest that this code may be duplicated and accounting for more of the binary
+than we would expect.
+
+FIXME: I'd like to have better measurements on the above!
+
+[salsa-rs/salsa#220]: https://github.com/salsa-rs/salsa/issues/220
+
+### Our goal
+
+The primary goal of this RFC is to make it so that the glue code we generate for
+query groups is not dependent on the database type, thus enabling better
+incremental rebuilds.
+
+## User's guide
+
+Most of the changes in this RFC are "under the hood". But there are various user
+visibile changes proposed here. 
```

Comment:
What happens to query definitions? 

Presumably, users now should always write them as `fn my_query(db: &dyn MyDatabase)`? Or, on the contrary, because we generate all the code in the context of query defining crate, it might makes sense to use `fn my_query(db: &iml MyDatabase)` as that would be monomorphised only once?

Reply:
Yes, they must write `dyn MyDatabase`. You could write `impl` but you'd have to write `impl MyDatabase + ?Sized`.